Add Kubeflow tensorboard viz and fix tensorflow file IO for cloud back-ends

stefannica opened the pull request:
## Describe changes
Enable the tensorboard visualization for Kubeflow and fixes the tensorflow file IO support when shared cloud storage like AWS S3 is used.

* fix the kubeflow container entrypoint to consider ZenML ModelArtifacts in addition to TFX ModelRun types when enabling the Tensorboard view
* when tensorflow is installed, TFX activates the tensorflow file IO plugin with a priority higher than the ZenML IO plugin. Use a higher priority for the ZenML file IO plugin
* the 2.6 tensorflow release moves the IO support into a separate python package `tensorflow_io` that must be installed and imported explicitly for the S3/GCP/Azure FS filesystem schemes to be supported
* fix the Kubeflow Metadata Store implementation
* suppress printing locals in rich trace logs if not in debug mode to reduce verbosity
* implement a TensorboardService to track and manage locally running Tensorboard daemons
* implement a step TensorboardVisualizer that starts a local Tensorboard daemon to visualize the entire history of a model logged by a step
* change the kubeflow example to use Tensorflow instead of scikit-learn and make use of Tensorboard visualizations


## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [ ] I have added tests to cover my changes.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

htahir1 reviewed (changes requested):
Thank you for taking this over Stefan! Really like it!

I would have some suggestions:

- Potentially adding this example as a file in the `kubeflow` example to showcase the tensorboard integration. I dont believe this deserves an extra example unless we have a native tensorboard integration that works in a magical way for different configurations
- Adding a picture of the 'Launch Tensorboard' button on the Kubeflow UI for users to understand the true value is when this is done on Kubeflow. This can be put on the `examples/kubeflow/README.md` in a subheading where we explain the kubeflow-tensorboard integration we just enabled
- Adding in the limitations of Kubeflow (e.g. with S3) as you decribed in the JIRA ticket
- Potentially adding a 'TensorboardVisualizer' and showcasing launching tensorboard in a post-execution workflow. This should be trivial but exemplrary

I can also implement some of the above if you give me the go-ahead. Otherwise, its its clear I would be glad to review these changes! Thanks again for doing the deep dive!

stefannica commented:
Thanks @htahir1 , I hope I've addressed all your suggestions with the last commit (see inline).

> Thank you for taking this over Stefan! Really like it!
> 
> I would have some suggestions:
> 
> * Potentially adding this example as a file in the `kubeflow` example to showcase the tensorboard integration. I dont believe this deserves an extra example unless we have a native tensorboard integration that works in a magical way for different configurations

I merged the Tensorboard and Kubeflow examples. This basically meant switching the existing Kubeflow example from sklearn to tensorflow and adding some post-execution visualizations.


> * Adding a picture of the 'Launch Tensorboard' button on the Kubeflow UI for users to understand the true value is when this is done on Kubeflow. This can be put on the `examples/kubeflow/README.md` in a subheading where we explain the kubeflow-tensorboard integration we just enabled

Done. There are new pics and new information regarding Tensorboard in that `README.md` file now.

> * Adding in the limitations of Kubeflow (e.g. with S3) as you decribed in the JIRA ticket

No need. I fixed all the limitations :)

> * Potentially adding a 'TensorboardVisualizer' and showcasing launching tensorboard in a post-execution workflow. This should be trivial but exemplrary

You're going to love this: I implemented a TensorboardService to wrap around the Tensorboard local server, and a visualizer that makes it easy to start/stop Tensorboard services. This goes along with the idea that Services are for more than just model serving. The visualizer also works with Jupyter notebooks, which reminds me: I also updated the notebook in the kubeflow example to reflect this.

schustmi commented on src/zenml/integrations/kubeflow/orchestrators/local_deployment_utils.py:
```diff
@@ -228,6 +228,129 @@ def deploy_kubeflow_pipelines(kubernetes_context: str) -> None:
+
+    Args:
+        kubernetes_context: The kubernetes context on which Kubeflow Pipelines
+            should be deployed.
```
```suggestion
            should be patched.
```

schustmi commented on src/zenml/integrations/tensorflow/visualizers/tensorboard_visualizer.py:
```diff
@@ -0,0 +1,216 @@
+        pipeline_name: the name of the pipeline
+        step_name: pipeline step name
+    """
+    repo = Repository()
+    pipeline = repo.get_pipeline(pipeline_name)
+    if pipeline is None:
+        raise RuntimeError(f"No pipeline with name `{pipeline_name}` was found")
+
+    last_run = pipeline.runs[-1]
+    step = last_run.get_step(name=step_name)
+    if step is None:
+        raise RuntimeError(
+            f"No pipeline step with name `{step_name}` was found in "
+            f"pipeline `{pipeline_name}`"
+        )
```
This can probably be extracted into a separate function as it's identical with the code above right?

schustmi commented on src/zenml/integrations/tensorflow/visualizers/tensorboard_visualizer.py:
```diff
@@ -0,0 +1,216 @@
+
+                # first check if a Tensorboard server is already running for
+                # the same logdir location and use that one
+                running_server = self.find_running_tensorboard_server(logdir)
+                if not running_server:
+                    return
+
+                logger.debug(
+                    "Stopping tensorboard server with PID '%d' ...",
+                    running_server.pid,
+                )
+                try:
+                    p = psutil.Process(running_server.pid)
+                except psutil.Error:
+                    logger.error(
+                        "Could not find process for PID '%d' ...",
+                        running_server.pid,
+                    )
+                    continue
+                p.kill()
```
Just for clarification: We need to implement it like this right now because ZenML doesn't track running services and the service isn't stored as an artifact somewhere right? Otherwise we could use `tensorboard_service.stop()` instead of killing the process manually?

stefannica replied:
Exactly ! Implementing extensible service registry and tracking is on my TODO list for the service abstraction. When that is done, ZenML should be able to load all running services on startup and this code will be replaced with a simple service registry lookup followed by a `service.stop()`.

schustmi reviewed (changes requested):
Nice! Just found a few typos and a tiny amount of duplicated code, other than that good to go

htahir1 reviewed (approved):
Are you kidding here!!!! Whatttt the heck just happened. AMAZING WORK! I absolutely love the new Tensorboard service and can only be astonished at the effort put in here. Kudos to you and LGTM!!!!

htahir1 reviewed (commented):
Ah wait  I just noticed one TINY problem. No updates to the docs. Would really love to have some of this stuff on the docs, including the service you implemented

htahir1 commented:
I have made the following changes to this branch.

* Added docs
* Added the following error to make sure kubeflow orchestrator does not fail non gracefully 
![image](https://user-images.githubusercontent.com/4792082/157670873-6a1bb305-16c2-4747-bf29-9da4dac117de.png)

@stefannica and @schustmi  please do take a look

stefannica commented on examples/kubeflow/README.md:
````diff
@@ -63,7 +72,7 @@ the model and the Tensorboard server will be updated automatically to include
 
 ```shell
-python run.py --learning_rate=0.02
+python run.py --lr=0.02
````
good catch, thanks !

stefannica reviewed (commented):
Nice job with the docs, I like it.

schustmi commented:
@htahir1 The only thing i'd change is to not use an `AssertionError`. Maybe a `RuntimeError` instead?
